### PR TITLE
Add architecture tuning with ENAS & hyperparameter tuning with parameter sharing

### LINKED ISSUE #50: Model developers to tune architecture
With ["Efficient Neural Architecture Search via Parameter Sharing"](https://arxiv.org/abs/1802.03268)

# Planned major changes

To better support architecture tuning with ENAS, I'm planning changes to Rafiki's current model training framework:

### Context

Currently, when application developers create model training jobs, they pass a budget like `{ 'GPU_COUNT': 1, 'MODEL_TRIAL_COUNT': 20 }`, with `MODEL_TRIAL_COUNT` deciding the no. of trials to conduct for *each model template*. 

### Change

Replace `MODEL_TRIAL_COUNT` option with `TIME_HOURS` option, which specifies how long the train job *should* run for. It is a soft time target. At the same time, I'll be reworking the Advisor component (which proposes trials' knobs) such that it is additionally in charge of deciding how many trials to run, when to stop each worker, when to stop the train job, given the budget e.g. `GPU_COUNT` and `TIME_HOURS`.

### Reasons for change
- May not be intuitive to the application developer to specify no. of trials while creating a train job ("how many trials should I put as budget? how long do I need to wait?"), especially if they're not supposed to be familiar with details like how model are trained and tuned. In contrast, `TIME_HOURS` is more straightforward.
- Currently, different models & model tuning strategies would require different no. of trials to be effective. For example, the original ENAS tuning strategy requires maybe (301x150+10+1) trials for sufficient train-eval cycles.
- In the future, it gives more flexibility for model tuning strategies at the Advisor component - for example, I'll be adding a new type of tuning strategy that takes all the models with no hyperparameters (e.g. model's knob config only consists of fixed values) and just conducts a *single* trial (since there's nothing to tune). It's also possible that a new tuning strategy can situationally conduct more/fewer trials based on feedback from workers.

### Motivation

I have been integrating ENAS as a new model tuning strategy on Rafiki (e.g. at the Advisor component). If model templates want to do architecture tuning with ENAS, the model's training code needs to switch between different "modes":

- During the ENAS architecture search phase, the model needs to *alternate* between "train my parameters for 1 epoch" and "don't train my parameters; just evaluate on the validation dataset"
- At the end of the architecture search, the model needs to switch to training its parameters from scratch with a *full-sized* architecture stacked with more cells, and train for 310 epochs

Similarly, when you think about a standard hyperparameter tuning procedure, you might want the model to do early-stopping for the first e.g. 100 trials, then conduct a final trial for a full e.g. 300 epochs. 

In both architecture tuning & hyperparameter tuning, the model needs to be configured by Rafiki somehow to switch between these "modes" on a trial-basis.

### Change

We can model the configuration of a model template for different training "modes" with different model *policies*. For example, if a model is to engage in policy `QUICK_TRAIN`, it is to prematurely speed up its training step e.g. by either doing early-stopping or reducing the no. of epochs. The model communicates to Rafiki which policies it supports by adding `PolicyKnob(policy_name)` to its `knob_config`. On the other hand, Rafiki configures the activation of the model's policies on a trial-basis by realising the values of `PolicyKnob`s to either `True` (activated) or `False` (not activated).

For example, here is a example knob config of a model which supports the policy `QUICK_TRAIN`:

image

Whenever the model is to do early-stopping, Rafiki will pass `quick_train=True` as part of the model's knobs. Otherwise, the model defaults to full-length training.

Here is my current documentation for `PolicyKnob`:
```python
'''
    Knob type representing whether a certain policy should be activated, as a boolean.
    E.g. the `QUICK_TRAIN` policy knob decides whether the model should stop model training early, or not. 
    Offering the ability to activate different policies can optimize hyperparameter search for your model. 
    Activation of all policies default to false.

    =====================       =====================
    **Policy**                  Description
    ---------------------       --------------------- 
    ``SHARE_PARAMS``            Whether model supports parameter sharing       
    ``QUICK_TRAIN``             Whether model should stop training early in `train()`, e.g. with use of early stopping or reduced no. of epochs
    ``SKIP_TRAIN``              Whether model should skip training its parameters
    ``QUICK_EVAL``              Whether model should stop evaluation early in `evaluate()`, e.g. by evaluating on only a subset of the validation dataset
    ``DOWNSCALE``               Whether a smaller version of the model should be constructed e.g. with fewer layers
    =====================       =====================
    
'''
```

### LINKED ISSUE #97: Model developers to share parameters across trials
Define model API for model developers to configure parameter sharing schemes and 
receive model parameters from other trials

### LINKED ISSUE #98: Add example image classification model that does architecture tuning
Refer to ENAS's child model at https://github.com/melodyguan/enas

Should replicate accuracy in paper

### LINKED ISSUE #99: Add example CIFAR-10 dataset-loading script

### LINKED ISSUE #101: Add novel parameter-sharing scheme
- [x] Experiment with multiple schemes and collate & compare results 
- [x] Implement best parameter-sharing scheme

### PR DESCRIPTION
**Pull request by nginyc**
Fixes #101 
Fixes #97 
Fixes #98 
Fixes #50 
Fixes #99 

## API Changes

For model developers:
- Parameters returned by `BaseModel.dump_parameters()` must now strictly conform to type `Dict[str, Union[str, int, float, np.ndarray]]`**
- Dataaset utils object `rafiki.model.dataset_utils` has been relocated into `rafiki.model.utils` as `utils.dataset`**
- Logger utils object `rafiki.model.logger` has been relocated into `rafiki.model.utils` as `utils.logger`**
- `test_model_class` has been relocated from `rafiki.model` into `rafiki.model.dev`**
- `rafiki.constants.TaskType` has been removed in favour of simply using strings (to not hard-code task names in core code)**
- New optional argument `shared_params` for `BaseModel.train()`**
- New optional static method `BaseModel.teardown()` to do any class-wide teardown logic across trials
- New `PolicyKnob` & `ArchKnob` knob types (to leverage on new cross-trial parameter sharing & architecture tuning features)

For application developers:
- New budget option `TIME_HOURS` for creating train jobs
- `rafiki.constants.BudgetType` has been renamed to `rafiki.constants.BudgetOption`**

For application users:
- Endpoint `/predict` for inference jobs now support batch predictions ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/quickstart-app-users.html#making-batch-predictions)) 

** specifies that change is NOT backward compatible

## Full Changelist

Major changes:
- Add budget option of `TIME_HOURS` ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/python/rafiki.client.html#rafiki.client.Client.create_train_job))
- Rework interface between advisor & train worker with the concept of "Proposals" for more flexible tuning schemes ([code](https://github.com/nginyc/rafiki/tree/add_arch_tuning/rafiki/advisor))
- Setup & document decision framework advisor to decide on which tuning scheme to use for each train job & model ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/model-tuning.html#model-tuning-schemes)) 
- Allow cross-trial parameter sharing with changes to the model API and leveraging on Redis for cross-worker sharing of parameters ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/python/rafiki.model.html#rafiki.model.BaseModel.train)]
- Add hyperparameter tuning with epsilon greedy parameter sharing as a tuning scheme, with a sample usage script `train_densenet.py` ([code](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/scripts/image_classification/train_densenet.py))
- Add ENAS as a tuning scheme, introducing `PolicyKnob` and `ArchKnob`, with  a sample usage script `run_enas.py` ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/model-tuning.html#architecture-tuning-with-enas), [code](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/scripts/image_classification/run_enas.py))
- Add making of batch predictions on inference jobs ([docs](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/quickstart-app-users.html#making-batch-predictions))
- Dynamically deploy a unique advisor for each model during training, instead of having a single static advisor across train jobs

Minor changes:
- Add example CIFAR-10 dataset loading script ([code](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/datasets/image_classification/load_cifar_format.py))
- Add sample script for converting directories of images to Rafiki's standard image classification format ([code](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/datasets/image_classification/load_folder_format.py))
- Add typing and clean up details in documentation of model interface & Rafiki Client
- Add & refactor abstract access layers for Redis & file system (to store datasets & model parameters)
- Move more responsibility of reading from the database from train worker to Advisor (prefer "smart" master, "dumb" worker)
- Rework service deployment to more accurately reflect statuses of train & inference jobs
- Add tests for inference job deployment

This branch's changes are largely reflected in as updated documentation deployed at https://nginyc.github.io/rafiki/docs/0.2.0. In particular, there is a new ["How Model Tuning Works" page](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/model-tuning.html), and we have 2 new model training scripts [`run_enas.py`](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/scripts/image_classification/run_enas.py) and [`train_densenet.py`](https://github.com/nginyc/rafiki/blob/add_arch_tuning/examples/scripts/image_classification/train_densenet.py).

Sorry for the long commit history - I was experimenting with ENAS for a while and there was a huge departure in code.

### CODE REVIEW

**Comment by nudles** [comment]
This is a big PR. Could you give a list of API changes for existing users? Would there be any issues if they want to run their existing examples after this PR is merged?

**Comment by nudles** [comment]
the documentation on hyper-parameter tuning is replaced by [model tuning](https://nginyc.github.io/rafiki/docs/0.2.0/src/user/model-tuning.html)? Model tuning doc is a bit complex. It would be better to start with simple schemes, e.g., just for hyper-parameter tuning. And then add advanced stuff, e.g., tuning policies like early stopping, and architecture tuning

**Comment by nginyc** [comment]
Hi @nudles, I have updated the details of the PR with API changes, some of which are not backward compatible. 

With regards to the new model tuning docs, there were no existing documentation on how model tuning is implemented on Rafiki. Following your feedback, I will be simplifying the new model tuning docs

**Comment by nginyc** [comment]
Have resolved the conflicts, improved the model tuning docs slightly, and did some testing. Ready to be merged
